core/src/storage/lazy.rs

```diff
@@ -0,0 +1,344 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+    KeyPtr,
+    Pull,
+    Push,
+    StorageSize,
+};
+use core::cell::UnsafeCell;
+use ink_primitives::Key;
+
+/// A lazy storage entity.
+///
+/// This loads its value from storage upon first use.
+///
+/// # Note
+///
+/// Use this if the storage field doesn't need to be loaded in some or most cases.
+#[derive(Debug)]
+pub struct Lazy<T> {
+    kind: UnsafeCell<LazyKind<T>>,
+}
+
+impl<T> StorageSize for Lazy<T>
+where
+    T: StorageSize,
+{
+    const SIZE: u64 = <T as StorageSize>::SIZE;
+}
+
+impl<T> Pull for Lazy<T>
+where
+    T: StorageSize + scale::Decode,
+{
+    fn pull(key_ptr: &mut KeyPtr) -> Self {
+        Self::lazy(key_ptr.next_for::<T>())
+    }
+}
+
+impl<T> Push for Lazy<T>
+where
+    T: Push,
+{
+    fn push(&self, key_ptr: &mut KeyPtr) {
+        // We skip pushing to contract storage if we are still in unloaded form.
+        if let LazyKind::Occupied(occupied) = self.kind() {
+            occupied.value.push(key_ptr)
+        }
+    }
+}
+
+impl<T> Lazy<T> {
+    /// Creates an eagerly populated lazy storage value.
+    #[must_use]
+    pub fn new(value: T) -> Self {
+        Self {
+            kind: UnsafeCell::new(LazyKind::Occupied(OccupiedLazy::new(value))),
+        }
+    }
+
+    /// Creates a true lazy storage value for the given key.
+    #[must_use]
+    pub fn lazy(key: Key) -> Self {
+        Self {
+            kind: UnsafeCell::new(LazyKind::Vacant(VacantLazy::new(key))),
+        }
+    }
+
+    /// Returns a shared reference to the inner lazy kind.
+    #[must_use]
+    fn kind(&self) -> &LazyKind<T> {
+        // SAFETY: We just return a shared reference while the method receiver
+        //         is a shared reference (&self) itself. So we respect normal
+        //         Rust rules.
+        unsafe { &*self.kind.get() }
+    }
+
+    /// Returns an exclusive reference to the inner lazy kind.
+    #[must_use]
+    fn kind_mut(&mut self) -> &mut LazyKind<T> {
+        // SAFETY: We just return an exclusive reference while the method receiver
+        //         is an exclusive reference (&mut self) itself. So we respect normal
+        //         Rust rules.
+        unsafe { &mut *self.kind.get() }
+    }
+
+    /// Performs the given closure on the mutable lazy kind.
+    ///
+    /// # Note
+    ///
+    /// Actions on the mutable lazy kind are performed within the closure
+    /// to not leak exclusive references to it to the outside. This is important
+    /// since the `for_kind` method itself operates only on `&self`.
+    fn for_kind<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&mut LazyKind<T>) -> R,
```

Comment:
An explicit lifetime annotation would be a good idea here, to ensure it's not setting the lifetime equal to `&'_ self`, which would allow escape.

```suggestion
        F: for<'a> FnOnce(&'a mut LazyKind<T>) -> R,
```